### ISSUE TITLE
Add Tailrec Annotation

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#905]
Add a `@tailrec` annotation like in Scala and verify that the function is tail recursive.

**Comment by mlutze** [comment]
Would the validation go in the TailRec compilation phase?

**Comment by magnus-madsen** [comment]
No- we have no checks after the Simplifier. It would probably go in phase after typing "close" to the redundancy phase. (It could even go in that phase, I guess?)

**Comment by magnus-madsen** [comment]
I guess it could go in the TailRec phase- but that would violate the principle that after the Simplifier we start to erase information from the AST.

**Comment by magnus-madsen** [comment]
OK; I guess putting it in TailRec makes sense-- I think the algorithm that is there pretty much matches what is required for the tailrec check.

**Comment by mlutze** [comment]
What do we think about the idea of inverting the annotation semantics, so that we need to annotate a non-tail-recursive function instead? That way, the user has to acknowledge a "dangerous" construction

**Comment by magnus-madsen** [comment]
> 
> 
> What do we think about the idea of inverting the annotation semantics, so that we need to annotate a non-tail-recursive function instead? That way, the user has to acknowledge a "dangerous" construction

I think this is a very interesting idea. What I wonder is how many functions will need to be marked as non-tail recursive. If it is the majority then its probably not a good idea. 

Sometime in the fall @racerand is planning to look at implementing tail recursion modulo cons. As part of that work, it might be interesting for him to investigate the distribution of tail recursive vs. non-tail recursive functions.

**Comment by magnus-madsen** [comment]
Wrong button

**Comment by magnus-madsen** [comment]
Something that I just realized: A function can be tail recursive and still cause a stackoverflow (i.e. by having a call before the tail call that overflows the stack). 

This means that the "inverse" of tail recursive is NOT "being stack unsafe". You can be stack unsafe even if you are tail recursive. 

Proving that a function is tail recursive or tail recursive modulo cons is easy, but proving that a call cannot overflow the stack is difficult because of higher-order functions and mutual recursion.

**Comment by mlutze** [comment]
Could stack safety be modeled as part of the type of a function, in a way similar to effects? That could address the higher-order function issue. Doesn't help the mutual recursion though, which I think may be in Halting Problem territory?

**Comment by magnus-madsen** [comment]
> 
> 
> Could stack safety be modeled as part of the type of a function, in a way similar to effects? That could address the higher-order function issue. Doesn't help the mutual recursion though, which I think may be in Halting Problem territory?

Yes, and yes. But it is probably to verbose to track stack safety in the type, but it is worth thinking about.

**Comment by magnus-madsen** [comment]
Steps:

- Add a `@Tailrec` annotation to `Weeder`.
- In `Safety` check a Boolean argument to `visitExp` to check that an expression is tail recursive.

**Comment by magnus-madsen** [comment]
@sanadKadu has expressed interest in working on this issue.

**Comment by magnus-madsen** [comment]
I want to rethink my approach to this.

**Comment by magnus-madsen** [comment]
Fixed in master 🎆
